gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,447 @@
+          DEFAULT_KAFKA_CLIENT_NAME);
+    }
+
+    @Override
+    protected long getEarliestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+      Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetRequestInfo =
+          Collections.singletonMap(new TopicAndPartition(partition.getTopicName(), partition.getId()),
+              new PartitionOffsetRequestInfo(kafka.api.OffsetRequest.EarliestTime(), 1));
+      return getOffset(partition, offsetRequestInfo);
+    }
+
+    @Override
+    protected long getLatestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException {
+      Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetRequestInfo =
+          Collections.singletonMap(new TopicAndPartition(partition.getTopicName(), partition.getId()),
+              new PartitionOffsetRequestInfo(kafka.api.OffsetRequest.LatestTime(), 1));
+      return getOffset(partition, offsetRequestInfo);
+    }
+
+    private long getOffset(KafkaPartition partition,
+        Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetRequestInfo)
+        throws KafkaOffsetRetrievalFailureException {
+      SimpleConsumer consumer =
+          this.getSimpleConsumer(partition.getLeader().getHost(), partition.getLeader().getPort());
+      for (int i = 0; i < NUM_TRIES_FETCH_OFFSET; i++) {
+        try {
+          OffsetResponse offsetResponse =
+              consumer.getOffsetsBefore(new OffsetRequest(offsetRequestInfo, kafka.api.OffsetRequest.CurrentVersion(),
+                  DEFAULT_KAFKA_CLIENT_NAME));
+          if (offsetResponse.hasError()) {
+            throw new RuntimeException("offsetReponse has error: "
+                + offsetResponse.errorCode(partition.getTopicName(), partition.getId()));
+          }
+          return offsetResponse.offsets(partition.getTopicName(), partition.getId())[0];
+        } catch (Exception e) {
+          LOG.warn(String.format("Fetching offset for topic %s, partition %d has failed %d time(s). Reason: %s",
+              partition.getTopicName(), partition.getId(), i + 1, e.getMessage()));
+          if (i < NUM_TRIES_FETCH_OFFSET - 1) {
+            try {
+              Thread.sleep((long) ((i + Math.random()) * 1000));
+            } catch (InterruptedException e2) {
+              LOG.error("Caught interrupted exception between retries of getting latest offsets. " + e2.getMessage());
+            }
+          }
+        }
+      }
+      throw new KafkaOffsetRetrievalFailureException(String.format(
+          "Fetching offset for topic %s, partition %d has failed.", partition.getTopicName(), partition.getId()));
+    }
+
+    @Override
+    protected Iterator<MessageAndOffset> fetchNextMessageBuffer(KafkaPartition partition, long nextOffset,
+        long maxOffset) {
+      if (nextOffset > maxOffset) {
+        return null;
+      }
+
+      FetchRequest fetchRequest = createFetchRequest(partition, nextOffset);
+      FetchResponse fetchResponse = null;
```

Comment:
This and line 324 can be combined since it is never used outside the try clause.